[DOC] Add pachctl mount docs (#4987)
pachyderm/pachyderm: doc/docs/master/how-tos/export-data-out-pachyderm/mount-repo-to-local-computer.md

> # Mount a Repo to a Local Computer

Reviewer: General note:

I think most of the information in this document is correct (see other comments for correctness improvement), but the sections should be reordered. 

Specifically, the default setting and use case for `pachctl mount` is mount all repos in read-only. We should start with that simple case and then talk about more complicated things afterwards, building up. I envision the sections to flow similar to:

Opening: pc mount lets you mount data in pach as a local directory. These mounts can be read-only when you're just looking to experiment or consume the data, or in read/write mode which lets you commit changes to the mounted data back into Pachyderm. 
1. Prereqs
2. Basic/default usage of pachctl mount (read-only)
3. How to mount only specific repos, branches, or commits (instead of everything)
4. Mounting data in read/write mode (use with caution see below)
5. Important note!! Avoiding merge conflicts.
6. A simple walkthrough of using writable mount with opencv. Maybe a separate doc/example and just link to it? I'm not sure which is better

Reviewer: Happy to discuss in TOH if you want. Again, in general the content is good, just need to fix the order the concepts are introduced

Author: so I think that prerequisites should go after. In the beginning, we discuss conceptual things - what is this feature, what are the use cases, limitations, etc. Then, we move on to the prerequisites and how to actually use it.

As for the example. What are you specifically looking for? There is a section below how to mount a Pachyderm repo. What else do you want to see there?

> 

Reviewer 2: We probably need a big red warning that fuse mounts are very unstable on mac at the moment

Sometimes you might need to access files directly in
a Pachyderm repository without downloading them first

> to your local computer. Pachyderm enables

Reviewer: This is not the use case for mount. Mount does not prevent you from needing to download the data to your local computer. It simply allows you to lazily download data -- meaning it only downloads the files when you try to access them. If you are accessing all the files, it's actually slower than just get file although it's convenient and users are often fairly familar with mounting file systems

you to do so with the `pachctl mount` command. This command

> uses the SSH Filesystem (SSHFS) client and Filesystem

Reviewer 2: Unless I'm missing something, I don't think we use sshfs anywhere

in Userspace (FUSE) user interface to export a Pachyderm
File System (PFS) to a Unix computer system. In other words,
it allows you to mount a Pachyderm repository on your computer.

For example, you have an [OpenCV pipeline](../../../getting_started/beginner_tutorial/#image-processing-with-opencv)
up and running,
and you need to edit files in the `images` repository. Let's
say you want to experiment with brightness and contrast
settings in `liberty.png`. Then, you want the pipelines
run for the updated file and see the result in the end.
If you do not mount the `images` repo, you would have to
first download the files to your computer, edit them,
and then put them back to the repository. The `pachctl mount`

> command automates all these steps for you. You can mount just the

Reviewer: None of this is technically incorrect, but it should come way later once we introduce the idea of the writable mount. I think once we re-order things a bit and turn it into the clearer narrative, we can come back and I can help edit this paragraph further although it probably doesnt need much work.

`images` repo or all Pachyderm repositories as directories
on you machine, edit as needed, and, when done,
exit the `pachctl mount` command. Upon exiting the `pachctl mount`
command, Pachyderm uploads all the changes to the corresponding
repository.

## Avoiding Merge Conflicts

> 

Reviewer: It needs to be made much more obvious (aka in their face) that using mount in write mode can be dangerous and cause complicated failures.  Mount in write mode should only be used when you have sole ownership over the mounted data. If this is not the case, merge conflicts or unexpected data overwrites can occur.

If someone else modifies the files while you are working on them
locally, their changes will likely be overwritten when you exit
`pachctl mount`. This happens because your changes are saved to
the Pachyderm repository only after you interrupt the `pachctl mount`

> command. Therefore, make sure that you do not work on the

Reviewer: A few questions on the functionality here:

1. Isn't there a pachctl unmount command (or umount)? You don't need to kill or ctrl+C the mount, although that does work of course. I've had a number of issues with the unmount cmd failing (on Mac) so might need to do more testing here. 
2. If you use a writable mount and mess up your data and want to just "throw it away" without actually writing back to the repo, is there a way to do that? 
3. Does umount or killing the process give the user any warning or prompt before writing back to the repo. I could see these mount being fairly long lived and users even forgetting their mounts are writable and then making unintended commits. This is more of a product question for @jdoliner perhaps, but are there any reasonable ways we can protect users more here?

Author: 1. there is no `pachctl unmount` command AFAIK. And it really just runs (block) in the terminal until you interrupt it. 
Of course, there are built-in UNIX commands that can mount/unmount your mounted dirs, which are literally `mount` and `unmount <disk>`. I don't know if we need to mention it in this doc. I presume it might be different on Ubuntu (`diskutil list` or similar).

2. Maybe @jdoliner  can answer this.
3. On macOS, none of this works. I don't think we have any protection against this

Reviewer 3: We do have a `pachctl unmount` command, it's just a wrapper around the `umount` syscall, so I don't think there's any difference between calling `umount /pfs` from the command line, and calling `pachctl unmount /pfs` (except of course that we spell it correctly). The only advantage of `pachctl unmount` is that it will unmount multiple filesystems is you don't pass it an argument. I'm not sure how compelling the use case is for that, it only really applies if you're mounting a bunch of filesystems, which seems dubious, and then you kinda lose track of them and want to unmount them all to cleanup. Honestly the biggest reason for that behavior is it was useful to me during dev because I often wound up with a lot of mounts from running tests.

same files while someone else is working on them.

Also, if you have any automated way
to upload files to a Pachyderm repo, and you are modifying files
on your local computer you might overwrite these changes with

> your changes, when you terminate the `pachctl mount` command.

Reviewer: This is the same idea as above. I think just combine the two. There's only one concept we're trying to get across -- if you use writable mount and the data in Pachyderm is changed (manually by another user or by an automated process), you can get merge conflicts and unexpected data overwrites.

> ## Access Permissions

Reviewer: This info is all good, but should be woven into a broader section introducing the use case for writable mounts. On it's own as an "access permissions" section doesn't really fit. It's much much more than just permissions, it's a whole major feature and use case (and new ways to shoot yourself in the foot -- with great power comes... users making more stupid mistakes too

`READ` access only. You can use the `--write` flag with the command
to enable `WRITE` access. However, even if you mount all the repos
with `WRITE` access, you cannot modify the files in the
Pachyderm output repositories. Because these repositories are
created by the Pachyderm pipelines, they are immutable. Only a pipeline
can change and update files in these repositories. If you try to change
a file in an output repo, you will get an error message.

## Mounting Branches and Commits

The `pachctl mount` command allows you to mount not only the default
branch, typically a `master` branch, but also other Pachyderm
branches. By default, Pachyderm mounts the `master` branch. However,
if you add a branch to the name of the repo, the `HEAD` of that branch
will be mounted. 

**Example:**

```bash
pachctl mount images --repos images@staging+w
```

You need to add the `+w` to the branch to enable write access.

Also, you can mount a specific commit. Because commits
might be on multiple branches, modifying them might result in data deletion
in the `HEAD` of the branches. That is why you can only mount commits in
read-only mode.

> ## Prerequisites

Reviewer: This section is great

Author: thanks!

You must have the following configured for this functionality to work:

* Unix or Unix-like operating system, such as Ubuntu 16.04 or macOS
Yosemite or later.
* FUSE and SSHFS for your operating system installed:

  * On macOS, run:

    ```bash
    brew cask install osxfuse
    ```

    ```bash
    brew install sshfs
    ```

  * On Ubuntu, run:

    ```bash
    sudo apt-get install sshfs
    ```

    For more information, see:

    * [FUSE for macOS](https://osxfuse.github.io/)

>     * [FUSE on Ubuntu](https://gist.github.com/cstroe/e83681e3510b43e3f618)

Reviewer 2: Why the link to this gist?

Reviewer 2: Definitely not the linked one, because it includes a bunch of errata that aren't relevant, like sshfs and zfs. Maybe the manpage? https://manpages.ubuntu.com/manpages/precise/man8/mount.fuse.8.html

Author: This page does not have information about how to actually install fuse on Ubuntu. I don't have Ubuntu to test, but I think `sudo apt-get install -y fuse` should do the trick. I'll remove the link.

  !!! note

>       macOS has limited support for FUSE and might not be stable.

Reviewer 2: No need to install sshfs

Author: removed

## Mount a Pachyderm Repo

Before you can mount a Pachyderm repo, verify that you have all the
[Prerequisites](#prerequisites).

To mount a Pachyderm repo on a local computer, complete the following
steps:

1. In a terminal, go to a directory in which you want your
Pachyderm repo. It can be any directory on your local computer,

> except for the `~/` directory. Mounting to the root directory will

Reviewer 2: This isn't true, it can't be any directory that's already mounted by another filesystem, which includes `~`. It has to be a new, empty directory.

Author: corrected.

   MkdirAll /var/folders/jl/mm3wrxqd75l9r1_d0zktphdw0000gn/T/pfs201409498/images
   ```

   The command runs in your terminal until you terminate it
   by pressing `CTRL+C`.

1. You can check that the repo was mounted by running the mount command
in your terminal:

   ```bash hl_lines="7"
   mount
   /dev/disk1s1 on / (apfs, local, read-only, journaled)
   devfs on /dev (devfs, local, nobrowse)
   /dev/disk1s2 on /System/Volumes/Data (apfs, local, journaled, nobrowse)
   /dev/disk1s5 on /private/var/vm (apfs, local, journaled, nobrowse)
   map auto_home on /System/Volumes/Data/home (autofs, automounted, nobrowse)
   pachctl@osxfuse0 on /Users/testuser/pachyderm/images (osxfuse, nodev, nosuid, synchronous, mounted by testuser)
   ```

1. Access your mountpoint.

   For example, in macOS, open Finder, press
   `CMD + SHIFT + G`, and type the mountpoint location. If you have mounted
   the repo to `~/pachyderm/images`, type `~/pachyderm/images`.

   ![finder-repo-mount](../../assets/images/s_finder_repo_mount.png)

1. Edit the files as needed.
1. When ready, add your changes to the Pachyderm repo by stopping

> the `pachctl mount` command with `CTRL+C`.

Reviewer: As noted in other comments -- does unmount work here? Is ti reliable on OSX?

Author: it is not reliable on macOS. Added pachctl unmount for reference.

# Mount a Repo to a Local Computer

!!! warning
    Pachyderm uses FUSE to mount repositories as local filesystems.
    Because Apple has announced phasing out support for macOS
    kernel extensions, including FUSE, this functionality is no
    longer stable on macOS Catalina (10.15) or later.

Pachyderm enables you to mount a repository
as a local filesystem on your computer by using the
`pachctl mount` command. This command

> uses Filesystem in Userspace (FUSE) user interface to export a Pachyderm

Reviewer 2: ```suggestion
uses the Filesystem in Userspace (FUSE) user interface to export a Pachyderm
```

# Mount a Repo to a Local Computer

!!! warning
    Because Pachyderm uses FUSE to mount a repo and
    macOS does not fully support FUSE, you might
    not be able to use this functionality on macOS.

Pachyderm enables you to mount a repository
as a local filesystem on your computer by using the
`pachctl mount` command. This command
uses Filesystem in Userspace (FUSE) user interface to export a Pachyderm
File System (PFS) to a Unix computer system.
This functionality is useful when you want to expose the results
of your pipeline for further consumption or modify the files

> in the input repository directly.

Author: added.

You can mount a Pachyderm repo in one of the following modes:

* `Read-only` — the users can view files in all
repositories, but cannot modify them.
* `Read-write` — the users can view files

> in all repositories and modify the files in input repositories.

Author: modified and added

## Mounting Repositories in Read-Only Mode

By default, when you run the `pachctl mount` command, Pachyderm mounts

> all Pachyderm repositories in read-only mode. You can access the

Reviewer 2: ```suggestion
all repositories in read-only mode. You can access the
```

Pachyderm sounds a bit repetitive otherwise

Author: corrected

files through your file browser or enable third-party applications
access. Read-only access enables you to explore and experiment with
the data, without modifying it. For example, you can mount your
repo to a local computer and then open that directory in a Jupyter
Notebook for exploration. 

## Mounting Specific Branches and Commits

The `pachctl mount` command allows you to mount not only the default
branch, typically a `master` branch, but also other Pachyderm
branches. By default, Pachyderm mounts the `master` branch. However,
if you add a branch to the name of the repo, the `HEAD` of that branch
will be mounted.

**Example:**

```bash

> pachctl mount images --repos images@staging+w

Author: corrected

```

You need to add the `+w` to the branch to enable write access.

Also, you can mount a specific commit. Because commits
might be on multiple branches, modifying them might result in data deletion
in the `HEAD` of the branches. That is why you can only mount commits in

> read-only mode.

Author: corrected

## Mounting Repositories in Read-Write Mode

Running the `pachctl mount` command with the `--write` flag grants you
write access to the mounted repositories, which means that you can

> open the files for editing and putting them back to the Pachyderm

Author: corrected.

repository. 

For example, you have the [OpenCV example pipeline](../../../getting_started/beginner_tutorial/#image-processing-with-opencv)
up and running,
and you need to edit files in the `images` repository. Let's
say you want to experiment with brightness and contrast
settings in `liberty.png`. Then, you want the pipelines
run for the updated file and see the result in the end.
If you do not mount the `images` repo, you would have to
first download the files to your computer, edit them,
and then put them back to the repository. The `pachctl mount`
command automates all these steps for you. You can mount just the
`images` repo or all Pachyderm repositories as directories
on you machine, edit as needed, and, when done,
exit the `pachctl mount` command. Upon exiting the `pachctl mount`
command, Pachyderm uploads all the changes to the corresponding

> repository.

Author: added.

    sudo apt-get install -y fuse
    ```

    For more information, see:

    * [FUSE for macOS](https://osxfuse.github.io/)

  !!! note
      macOS has limited support for FUSE and might not be stable.

## Mount a Pachyderm Repo

Before you can mount a Pachyderm repo, verify that you have all the
[Prerequisites](#prerequisites).

To mount a Pachyderm repo on a local computer, complete the following
steps:

1. In a terminal, go to a directory in which you want to mount a
Pachyderm repo. It can be any new empty directory on your local computer.
For example, `pfs`.

1. Run `pachctl mount` for a repository and branch that you want to mount:

   ```bash
   pachctl mount <path-on-your-computer> [flags]
   ```

   **Example:**

   * If you want to mount all the repositories in your Pachyderm cluster 
   to a `pfs` directory on your computer and give `WRITE` access to them, run:

   ```bash
   pachctl mount pfs --write
   ```

   * If you want to mount the master branch of the `images` repo
   and enable file editing in this repository, run:

   ```bash
   pachctl mount pfs --repos images@master+w
   ```

   To give read-only access, omit `+w`.

   **System Response:**

   ```

>    ro for images: &{Branch:master Write:true}

Reviewer: We might need someone with linux to try running through this and write the system responses. NO point in having all the mac-specific stuff here

access. Read-only access enables you to explore and experiment with
the data, without modifying it. For example, you can mount your
repo to a local computer and then open that directory in a Jupyter
Notebook for exploration. 

## Mounting Specific Branches and Commits

The `pachctl mount` command allows you to mount not only the default
branch, typically a `master` branch, but also other Pachyderm
branches. By default, Pachyderm mounts the `master` branch. However,
if you add a branch to the name of the repo, the `HEAD` of that branch
will be mounted.

**Example:**

```bash
pachctl mount images --repos images@staging+w
```

You need to add the `+w` to the branch to enable write access.

Also, you can mount a specific commit. Because commits
might be on multiple branches, modifying them might result in data deletion
in the `HEAD` of the branches. That is why you can only mount commits in
read-only mode.

## Mounting Repositories in Read-Write Mode

Running the `pachctl mount` command with the `--write` flag grants you
write access to the mounted repositories, which means that you can
open the files for editing and putting them back to the Pachyderm
repository. 

> For example, you have an [OpenCV pipeline](../../../getting_started/beginner_tutorial/#image-processing-with-opencv)

Reviewer 2: ```suggestion
For example, you have the [OpenCV example](../../../getting_started/beginner_tutorial/#image-processing-with-opencv)
```

command, Pachyderm uploads all the changes to the corresponding
repository.

If someone else modifies the files while you are working on them
locally, their changes will likely be overwritten when you exit
`pachctl mount`. This happens because your changes are saved to
the Pachyderm repository only after you interrupt the `pachctl mount`
command. Therefore, make sure that you do not work on the
same files while someone else is working on them.

!!! warning
    Use writable mount **ONLY** when you have sole ownership
    over the mounted data. Otherwise, merge conflicts or
    unexpected data overwrites can occur.

Because output repositories are created by the Pachyderm
pipelines, they are immutable. Only a pipeline
can change and update files in these repositories. If you try to change
a file in an output repo, you will get an error message.

## Prerequisites

You must have the following configured for this functionality to work:

* Unix or Unix-like operating system, such as Ubuntu 16.04 or macOS
Yosemite or later.
* FUSE for your operating system installed:

  * On macOS, run:

    ```bash
    brew cask install osxfuse
    ```

    ```bash
    brew install sshfs

>     ```

Reviewer 2: this still needs to be removed

  * On Ubuntu, run:

    ```bash
    sudo apt-get install sshfs

>     ```

Reviewer 2: still need to remove this

Reviewer 2: I'm not sure if the diff changed or I misread, but the comment I left was for a `brew install sshfs` instruction. Generally I think it makes sense to keep the macOS instructions even if our fuse support is pretty poor. We'll hopefully fix that in the future.

Author: @ysimonson I thought I did remove just `brew install sshfs` only. The macOS instructions are above the Ubuntu instructions. Is something missing there?

Reviewer 2: Nope, I'm clarifying that the first comment in this thread is no longer relevant

    * [FUSE for macOS](https://osxfuse.github.io/)
    * [FUSE on Ubuntu](https://gist.github.com/cstroe/e83681e3510b43e3f618)

  !!! note
      macOS has limited support for FUSE and might not be stable.

## Mount a Pachyderm Repo

Before you can mount a Pachyderm repo, verify that you have all the
[Prerequisites](#prerequisites).

To mount a Pachyderm repo on a local computer, complete the following
steps:

1. In a terminal, go to a directory in which you want to mount a
Pachyderm repo. It can be any new empty directory on your local computer.
For example, `pfs`.

1. Run `pachctl mount` for a repository and branch that you want to mount:

   ```bash
   pachctl mount <path-on-your-computer> [flags]
   ```

   **Example:**

   * If you want to mount all the repositories in your Pachyderm cluster 
   to a `pfs` directory on your computer and give `WRITE` access to them, run:

   ```bash
   pachctl mount pfs --write
   ```

   * If you want to mount the master branch of the `images` repo
   and enable file editing in this repository, run:

   ```bash
   pachctl mount images --repos images@master+w
   ```

   To give read-only access, omit `+w`.

   **System Response:**

>    ```bash

Reviewer 2: terminal output should not be bash highlighted

   continue
   ri: repo:<name:"images" > created:<seconds:1591812554 nanos:28450609 > size_bytes:244068 branches:<repo:<name:"images" > name:"master" >
   MkdirAll /var/folders/jl/mm3wrxqd75l9r1_d0zktphdw0000gn/T/pfs201409498/images
   ```

   The command runs in your terminal until you terminate it
   by pressing `CTRL+C`.

1. You can check that the repo was mounted by running the mount command
in your terminal:

   ```bash hl_lines="7"
   mount
   /dev/disk1s1 on / (apfs, local, read-only, journaled)
   devfs on /dev (devfs, local, nobrowse)
   /dev/disk1s2 on /System/Volumes/Data (apfs, local, journaled, nobrowse)
   /dev/disk1s5 on /private/var/vm (apfs, local, journaled, nobrowse)
   map auto_home on /System/Volumes/Data/home (autofs, automounted, nobrowse)
   pachctl@osxfuse0 on /Users/testuser/pachyderm/images (osxfuse, nodev, nosuid, synchronous, mounted by testuser)
   ```

1. Access your mountpoint.

   For example, in macOS, open Finder, press
   `CMD + SHIFT + G`, and type the mountpoint location. If you have mounted

>    the repo to `~/pachyderm/images`, type `~/pachyderm/images`.

Reviewer 2: Currently we're referencing, as example directories, `pfs` above, and `~/pachyderm/images` here. How about using the same example directory across the board?
